### PR TITLE
Added feature "If DokuWiki page has no title then it needs to be shown in the admin interface"

### LINKED ISSUE #12: If DokuWiki page has no title then it needs to be shown in the admin interface

### LINKED ISSUE #28: Exported IDs should be unique
# Bug

## Reproduce

1. Call first export page, eg. `count=1&start=0`.
2. Call second export page, eg. `count=1&start=1`.

## What should happen

* The item in the first export page should have ID 0.
* The item in the second export page should have ID 1.

## What happens

* The item in the first export page has ID 0.
* The item in the second export page has ID 0.

### PR DESCRIPTION
**Pull request by TheKeymaster**
Resolve #12 
Fix #28

### CODE REVIEW

**Comment by TheKeymaster** [comment]
@georgms could you please review this?

File: admin.php [hunk 1]
```diff
@@ -7,39 +7,148 @@
...
     }
 
     /**
-     * @return bool true if only access for superuser, false is for superusers and moderators
+     * Admins or higher can access this plugin.
+     * It only has page data that is accessible for admins.
+     * You do not need to be a superuser to access this plugin.
      */
-    public function forAdminOnly() {
+    public function forAdminOnly()
```
Comment (georgms) [inline admin.php]: Why is it for admins only? Every user could potentially add missing titles.

File: admin.php [hunk 2]
```diff
@@ -7,39 +7,148 @@
...
+            $url[] = wl($page, '', true);
+
+            $timeModified = new DateTime('@' . $metadata[$key]['last_change']['date']);
+            $timeModified->add(new DateInterval('PT2H'));
+            $modifiedTimeStamp[] = $timeModified->format(self::TIME_FORMAT) . ' ' . self::TIMEZONE; // Last edit
+        }
+
+        // Put variables to array
+        $variables = array(
+            'pagesWithoutTitle' => $pagesWithoutTitle,
```
Comment (georgms) [inline admin.php]: Use constants instead of magic strings.

File: admin.php [hunk 3]
```diff
@@ -7,39 +7,148 @@
...
+            );
+
+        return $variables;
+    }
+
+    private function getTranslations() {
+
+        // Get translations from DokuWiki
+        $translations = array(
+            'menu' => $this->getLang('menu'),
```
Comment (georgms) [inline admin.php]: Use constants instead of magic strings.

Comment (georgms) [reply admin.php]: What's the point of this at all? The respective `lang.php` already builds a `$lang` variable with all translations. Try to pass that to the template and use it there.

**Review by georgms — CHANGES_REQUESTED** [review]
* Needs more tests!

File: admin.php (hunk 1 above)
Comment (TheKeymaster) [reply admin.php]: This method says like: 
*Do you want to enable this plugin for Admins or for Superadmins only?*

I guess the best way is to also show this for *Admins* as well.

File: admin.php (hunk 3 above)
Comment (TheKeymaster) [reply admin.php]: Add array to template

File: admin.php (hunk 2 above)
Comment (TheKeymaster) [reply admin.php]: ok

**Comment by TheKeymaster** [comment]
Note4Me:

Make a new class for the pagesWithNoTitle method and make sure to add tests for the html() function with output buffering.

**Comment by TheKeymaster** [comment]
@georgms do we need more test cases? Could you review this?

**Comment by TheKeymaster** [comment]
This PR also fixes #28

File: PageGetter.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<?php
+/**
+ * This is the Dokuwiki export made by Dominik Brader for FINDOLOGIC.
```
Comment (georgms) [inline PageGetter.php]: Author info is already included in `@author`.

File: PageGetter.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * @return array all pages that do not have a title set.
+     */
+    static function getPagesWithoutTitle()
+    {
+        $indexer = new Doku_Indexer();
+        $allPages = $indexer->getPages();
+
+        foreach ($allPages as $page) {
+            if (p_get_metadata($page)['description']) { // Only get pages with content
```
Comment (georgms) [inline PageGetter.php]: What about pages without description and without title? The comment should not only describe that you skip them, but especially also why you skip them.

Comment (georgms) [reply PageGetter.php]: Instead of the `foreach` loop you could use `array_filter` with a nice closure, but it's not necessarily better.

File: DokuwikiXMLExport.php [hunk 6]
```diff
@@ -129,7 +129,7 @@ public function generateXMLExport($start, $count)
 
         $items = array();
         foreach ($this->pages as $key => $page) {
-            $item = $exporter->createItem($key);
+            $item = $exporter->createItem($start + $key);
```
Comment (georgms) [inline DokuwikiXMLExport.php]: Why did you change that? If it was a bug write a test for it.

File: _test/page.getter.test.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+//======================================================================
+// PAGEGETTER TESTS
+//======================================================================
+
+require_once(__DIR__ . '/../DokuwikiXMLExport.php');
+require_once(__DIR__ . '/../admin.php');
+require_once(__DIR__ . '/../_test/Helper.php');
+require_once(__DIR__ . '/../PageGetter.php');
+
+class page_getter_test extends DokuWikiTest
```
Comment (georgms) [inline _test/page.getter.test.php:11]: It is good practice that the filename and the class name should match, for example this will allow PHP to automatically discover in which file a class is to be found.

File: _test/page.getter.test.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public function setUp()
+    {
+        Helper::setUp();
+    }
+    public function test_get_pages_without_title_only_gets_pages_without_title()
```
Comment (georgms) [inline _test/page.getter.test.php]: Test

* 0: Empty Wiki; Wiki where all pages have a title.
* 1: The test covers this already.
* Multiple: Multiple pages without a title.
* A lot: What if you have 10.000.000 pages without a title? Does this work? How does your UI handle this?

File: _test/template.response.test.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        // Call html() function from the plugin
+        $adminPlugin = new admin_plugin_findologicxmlexport();
+        $adminPlugin->html();
+        $output = ob_get_clean();
+        $xml = new SimpleXMLElement($output);
+
+        $header = (string)$xml->xpath('h1')[0];
+        $expectedHeader = 'FINDOLOGIC XML Export Plugin';
+
+        $successInformation = (string)$xml->xpath('fieldset/div')[0];
```
Comment (georgms) [inline _test/template.response.test.php]: Maybe `trim()` the output so you don't have to care about that in the `$expected` text.

File: _test/template.response.test.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        // Output Buffer
+        ob_start();
+
+        // Call html() function from the plugin
+        $adminPlugin = new admin_plugin_findologicxmlexport();
+        $adminPlugin->html();
+        $output = ob_get_clean();
+        $xml = new SimpleXMLElement($output);
+
+        $header = (string)$xml->xpath('h1')[0];
```
Comment (georgms) [inline _test/template.response.test.php]: It's good practice to add a class or ID to elements. This will make the HTML output more semantic, and the test more robust since you can address the element using the class / ID.

File: _test/template.response.test.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        $legend = (string)$xml->xpath('fieldset/legend')[0];
+        $expectedLegend = 'Pages without title (0)';
+
+        $this->assertEquals($expectedHeader, $header, 'Expected header should be equal to the translation set in the english translation file.');
+        $this->assertEquals($expectedSuccessInformation, $successInformation, 'Expected success information should be equal to the translation set in the english translation file.');
+        $this->assertEquals($expectedLegend, $legend, 'Expected legend should be equal to the translation set in the english translation file.');
+
+    }
+
+    public function test_response_has_one_element_when_one_page_has_no_title_set()
```
Comment (georgms) [inline _test/template.response.test.php:54]: Are the pages without title displayed in the output? You should test that they are.

File: _test/template.response.test.php [hunk 12]
```diff
@@ -0,0 +1,85 @@
+<?php
+//======================================================================
+// TEMPLATE RESPONSE TESTS
+//======================================================================
+
+require_once(__DIR__ . '/../DokuwikiXMLExport.php');
+require_once(__DIR__ . '/../admin.php');
+require_once(__DIR__ . '/../_test/Helper.php');
+
+class template_response_test extends DokuWikiTest
```
Comment (georgms) [inline _test/template.response.test.php:10]: Test 0, 1, multiple, a lot.

Comment (georgms) [reply _test/template.response.test.php:10]: Test Umlaute, special characters etc. do not mess up your output.

File: admin.php [hunk 13]
```diff
@@ -7,39 +7,124 @@
...
-    define('DOKU_INC',realpath(dirname(__FILE__).'/../../../').'/');
+if (!defined('DOKU_INC')) {
+    die('Must be run within Dokuwiki!');
 }
 
-class admin_plugin_findologicxmlexport extends DokuWiki_Admin_Plugin {
+require_once(__DIR__ . '/vendor/autoload.php');
+require_once(__DIR__ . '/PageGetter.php');
+
+class admin_plugin_findologicxmlexport extends DokuWiki_Admin_Plugin
```
Comment (georgms) [inline admin.php]: Conventially class names start with a capital letter. Ideally they should use caps case, but Dokuwiki apparently uses some weird underscore caps case so you should stick to that.

File: admin.php [hunk 14]
```diff
@@ -7,39 +7,124 @@
...
+     * The lower this value is, the higher it is sorted.
      */
-    public function getMenuSort() {
-        return 10000;
+    const MENU_SORT = 1;
+
+    /**
+     * This is the time format as it gets outputted to the template.
+     */
+    const TIME_FORMAT = 'd.F Y (H:i)';
```
Comment (georgms) [inline admin.php]: Can you localize that?

File: admin.php [hunk 15]
```diff
@@ -7,39 +7,124 @@
...
     }
 
     /**
-     * Should carry out any processing required by the plugin.
+     * HTML output (gets generated by twig).
      */
-    public function handle() {
+    public function html()
+    {
+        global $conf;
```
Comment (georgms) [inline admin.php]: Ugh, global. But if this is how Dokuwiki does things I guess it's fine.

File: admin.php [hunk 16]
```diff
@@ -7,39 +7,124 @@
...
+        $pagesWithoutTitle = PageGetter::getPagesWithoutTitle();
+        $variables = $this->getVariablesForTemplate($pagesWithoutTitle);
+
+        $loader = new Twig_Loader_Filesystem(self::TEMPLATE_DIR);
+        $twig = new Twig_Environment($loader);
+
+        $initLanguage = $this->getLang('');
```
Comment (georgms) [inline admin.php]: What happens if you pass an empty string to `getLang()`. Please comment.

File: admin.php [hunk 17]
```diff
@@ -7,39 +7,124 @@
...
+
+        $twig->addGlobal('languageText', $this->lang);
```
Comment (georgms) [inline admin.php]: Does this have to be global?

File: admin.php [hunk 18]
```diff
@@ -7,39 +7,124 @@
...
+        );
+
+        return $variables;
+    }
+
+    /**
+     * @param $unixTime int from now.
+     * @return string Time as Formatted string with timezone.
+     */
+
```
Comment (georgms) [inline admin.php]: Docblock should stick to method declaration.

File: admin.php [hunk 19]
```diff
@@ -7,39 +7,124 @@
...
+            'urls' => $url,
+            'timestamp' => $modifiedTimeStamp,
+            'imageUrl' => self::EDIT_IMAGE_URL,
+        );
+
+        return $variables;
+    }
+
+    /**
+     * @param $unixTime int from now.
```
Comment (georgms) [inline admin.php]: `int from now`? What?

File: admin.php [hunk 20]
```diff
@@ -7,39 +7,124 @@
...
+    private function getVariablesForTemplate($pagesWithoutTitle)
+    {
+
+        // Generate variables based on page data
+        foreach ($pagesWithoutTitle as $key => $page) {
+            $metadata[] = p_get_metadata($page);
+            $url[] = wl($page, '', true);
+
+            // Make new DateTime Object and Format it
+            $modifiedTimeStamp[] = $this->formatTime($metadata[$key]['last_change']['date']);
```
Comment (georgms) [inline admin.php]: I would refactor this so that `PageGetter` returns an array of objects containing all required information - page, author, timestamp, whatever. This will make the data more cohesive and keep business logic out of the controller.

You should declare a class for this data and not just put them in a map like `['page' => ..., 'author' => ..., ...]`. This will give you some type safety and makes it easier to refactor the code etc.

File: admin.php [hunk 21]
```diff
@@ -7,39 +7,124 @@
...
+    }
+
+    /**
+     * @param $unixTime int from now.
+     * @return string Time as Formatted string with timezone.
+     */
+
+    private function formatTime($unixTime) {
+        $time = new DateTime('@' . $unixTime);
+        $timeInCest = $time->add(new DateInterval('PT2H'));
```
Comment (georgms) [inline admin.php]: If I understand this correct you are adding 2 hours to the UNIX timestamp to get the CEST time. This is totally right... for half a year. I am sure there is a better way to do this.

It's good that you are adding the timezone to the time, this avoids confusion. I would however prefer to see the time in my local timezone.

File: tpl/admin.tpl [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+                        {{languageText['lastedited']}}
+                    </th>
+                    <th>
+                        {{languageText['edit']}}
+                    </th>
+                </tr>
+                {% set amount = 0 %}
+                {% set morePages = 0 %}
+                {% for pageWithoutTitle in pagesWithoutTitle %}
+                {% if amount < 5 %}
```
Comment (georgms) [inline tpl/admin.tpl]: The controller should determine how many pages are displayed before they are cut off. Keep the template as stupid as possible, it should just concern itself with building HTML.

File: tpl/admin.tpl [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+                        {# edit #}
+                        <a target="_blank" class='editpage' style="color: transparent;" href="{{urls[amount]}}&do=edit">
+                            <div class='editpage'
+                                 style='background-image:url("{{imageUrl}}");background-repeat: no-repeat;background-size:25px;width: 25px;height: 25px;color: transparent; font-size: 20px;'>
+                            </div>
+                        </a>
+                    </td>
+                    {% set amount = amount + 1 %}
+                    {% else %}
+                    {% set morePages = morePages + 1 %}
```
Comment (georgms) [inline tpl/admin.tpl]: Also, should be provided by the controller.

**Review by georgms — CHANGES_REQUESTED** [review]
* Nice code. Lots of constants, comments and nice, lean methods with a strict purpose. Well done!
* There are a lot of comments, but it's also a lot of code. You're doing fine.
* Once you're done please provide admin page screenshots of the output for 0, 1, >5 pages without title.

File: DokuwikiXMLExport.php (hunk 6 above)
Comment (TheKeymaster) [reply DokuwikiXMLExport.php]: Totally agreed, forgot about it.

File: _test/template.response.test.php (hunk 12 above)
Comment (TheKeymaster) [reply _test/template.response.test.php:10]: > Test Umlaute, special characters etc. do not mess up your output.

DokuWiki does not allow pages with umlauts or special characters. All special characters are set to `_` and all umlauts are set to their corresponding form e.g. `ä = ae`.

File: admin.php (hunk 14 above)
Comment (TheKeymaster) [reply admin.php]: Yes I can.

File: admin.php (hunk 15 above)
Comment (TheKeymaster) [reply admin.php]: Well to be honest, this global isn't even used, so I will gladly remove it :sweat_smile:.

File: admin.php (hunk 16 above)
Comment (TheKeymaster) [reply admin.php]: Totally.

File: admin.php (hunk 17 above)
Comment (TheKeymaster) [reply admin.php]: Yes because in Twig I am only able to set variables globally or in the templatefile itself. So I chose this way.

File: admin.php (hunk 19 above)
Comment (TheKeymaster) [reply admin.php]: haha.. I meant the Unix timestamp.

File: admin.php (hunk 18 above)
Comment (TheKeymaster) [reply admin.php]: Thanks for noticing!

File: admin.php (hunk 21 above)
Comment (TheKeymaster) [reply admin.php]: Using the local timezone now..

File: tpl/admin.tpl (hunk 23 above)
Comment (TheKeymaster) [reply tpl/admin.tpl]: I am not sure how I should do this.
Could we discuss that soon-ish?

File: admin.php (hunk 20 above)
Comment (TheKeymaster) [reply admin.php]: Totally, that would be awesome!

File: admin.php (hunk 13 above)
Comment (TheKeymaster) [reply admin.php]: DokuWiki weird Approved &copy;

File: _test/template.response.test.php (hunk 11 above)
Comment (TheKeymaster) [reply _test/template.response.test.php:54]: Yes they are. I will implement tests for it!

File: _test/template.response.test.php (hunk 9 above)
Comment (TheKeymaster) [reply _test/template.response.test.php]: Aahhh I knew there was a solution but I wasn't sure how. Thanks for this!

File: tpl/admin.tpl (hunk 23 above)
Comment (georgms) [reply tpl/admin.tpl]: All the values calculated in the `{% set ... %}` statements should be provided by the controller. So that's about:

* An array of the pages that should be listed (ie. containing max. 5 elements).
* The number of omitted pages, if any.

Then the sort of looks like this:

```
for each $page in $pagesToDisplay
  display $page
if $morePages > 0
  show hint that there are $morePages pages
```

**Review by georgms — CHANGES_REQUESTED** [review]

File: _test/template.response.test.php (hunk 12 above)
Comment (georgms) [reply _test/template.response.test.php:10]: Keep logic as much as possible out of the template. This will make it easier to test and make the responsibilities of the different components much clearer.

File: admin.php [hunk 24]
```diff
@@ -1,45 +1,131 @@
...
+    {
+        $pagesWithoutTitle = PageGetter::getPagesWithoutTitle();
+
+        $urls = [
+            'editImageUrl' => self::EDIT_IMAGE_URL,
+            'exportUrl' => self::EXPORT_URL,
+            'stylesheetUrl' => self::STYLESHEET_URL
+        ];
+
+        $variablesForTemplate = ['var' => array_merge($pagesWithoutTitle, $urls)];
```
Comment (georgms) [inline admin.php]: Put the variables at top level, not within `var`.

File: admin.php [hunk 25]
```diff
@@ -1,45 +1,131 @@
...
+
+        // Set up loader and environment for twig.
+        $loader = new Twig_Loader_Filesystem(self::TEMPLATE_DIR);
+        $twig = new Twig_Environment($loader);
+
+        $initLanguage = $this->getLang(''); // Loads the translations set in the plugin translation.
+
+        // Add globals because this is the only way to add variables without declaring them directly in the template.
+        $twig->addGlobal(self::LANGUAGE_VARIABLE_NAME, $this->lang);
```
Comment (georgms) [inline admin.php]: Add the translations to the `$variablesForTemplate`, makes sense to have them in a subkey.

File: tpl/admin.tpl [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+                        {{languageText['lasteditby']}}
+                    </th>
+                    <th class="header-page-last-edited">
+                        {{languageText['lastedited']}}
+                    </th>
+                    <th class="header-page-edit-link">
+                        {{languageText['edit']}}
+                    </th>
+                </tr>
+                {% for page in page..4 %} {# 5 pages are shown #}
```
Comment (georgms) [inline tpl/admin.tpl]: This should just loop over a number of pages received from the controller and output all of them. The template should not decide how many pages to display.

File: tpl/admin.tpl [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+                        {{var[page]['lastEdit']}}
+                    </td>
+                    <td class="page-edit-link">
+                        {# edit #}
+                        <a target="_blank" class='page-edit-link-a' href="{{var[page]['url']}}&do=edit">
+                            <div class='page-edit-link-div' style='background-image:url("{{var['editImageUrl']}}");'></div>
+                        </a>
+                    </td>
+                </tr>
+                {% elseif page == 4 %}
```
Comment (georgms) [inline tpl/admin.tpl]: The controller should pass a variable like `pagesSkipped`.

File: admin.php [hunk 28]
```diff
@@ -1,45 +1,131 @@
...
+            'stylesheetUrl' => self::STYLESHEET_URL
+        ];
+
+        $variablesForTemplate = ['var' => array_merge($pagesWithoutTitle, $urls)];
+
+        // Set up loader and environment for twig.
+        $loader = new Twig_Loader_Filesystem(self::TEMPLATE_DIR);
+        $twig = new Twig_Environment($loader);
+
+        $initLanguage = $this->getLang(''); // Loads the translations set in the plugin translation.
```
Comment (georgms) [inline admin.php]: * You can skip the return variable.
* Comment more on this (something about `getLang()` needs to be called once to initialize `$this->lang`).

File: PageGetter.php [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+        }, ARRAY_FILTER_USE_BOTH);
+
+        // Amount of pages and edit button image url.
+        $pagesData['amount'] = count($allPagesWithoutTitle);
+
+        $sortedPages = array_values($allPagesWithoutTitle); // Sort pages so they start with array key [0]
+
+        foreach ($sortedPages as $key => $sortedPage) {
+            $dokuWikiPage = new DokuwikiPage($sortedPage);
+            $pagesData[$key] = get_object_vars($dokuWikiPage);
```
Comment (georgms) [inline PageGetter.php]: Converting from object to array should not be required, Twig should be able to handle objects just fine.

File: DokuwikiPage.php [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+    public $id;
+    public $url;
+    public $author;
+    public $lastEdit;
+    public $metadata;
+
+    /**
+     * Gets DokuWiki page data.
+     *
+     * * **id** - ID.
```
Comment (georgms) [inline DokuwikiPage.php]: Add field explanations to the fields as docblocks.

**Review by georgms — CHANGES_REQUESTED** [review]
* Take care of comments.
* Tests, Tests, Tests!

File: DokuwikiPage.php (hunk 30 above)
Comment (TheKeymaster) [reply DokuwikiPage.php]: :white_check_mark: done.

File: PageGetter.php (hunk 29 above)
Comment (TheKeymaster) [reply PageGetter.php]: :white_check_mark:

File: admin.php (hunk 24 above)
Comment (TheKeymaster) [reply admin.php]: :white_check_mark:

File: admin.php (hunk 28 above)
Comment (TheKeymaster) [reply admin.php]: :white_check_mark:

File: admin.php (hunk 25 above)
Comment (TheKeymaster) [reply admin.php]: :white_check_mark:

File: tpl/admin.tpl (hunk 26 above)
Comment (TheKeymaster) [reply tpl/admin.tpl]: :white_check_mark:

File: tpl/admin.tpl (hunk 27 above)
Comment (TheKeymaster) [reply tpl/admin.tpl]: :white_check_mark:

**Comment by TheKeymaster** [comment]
## TODOS

In `\page_getter_test` add following Tests:

TESTS:
 * 0: Empty Wiki; Wiki where all pages have a title.
 * 1: The test covers this already.
 * Multiple: Multiple pages without a title.
 * A lot: What if you have 10.000.000 pages without a title? Does this work? How does your UI handle this?

**Comment by TheKeymaster** [comment]
> Once you're done please provide admin page screenshots of the output for 0, 1, >5 pages without title.

Just because it is quick and dirty here are some screenshots of the backend (German):

 * 0:
selection_135
 * 1:
selection_134


 * .>5:
selection_133

**Comment by georgms** [comment]
Add tooltip explaining why pages without title are bad.

File: tpl/admin.tpl [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+        {% if totalPages > 0 %}
+        <div id="fl-notify-warning" class="notify">
+            {{languageText['noTitleWarning']}} <br/> {{languageText['noTitleWarningMoreInformation']}}
+        </div>
+        {% else %}
+        <div id="fl-success" class="success">
+            {{languageText['allPagesHaveATitle']}}
+        </div>
+        {% endif %}
+        <legend id="fl-legend">{{languageText['pagesWithoutTitle']}} ({{totalPages}})</legend>
```
Comment (georgms) [inline tpl/admin.tpl]: Add tooltip explaining why pages without title are bad.

File: DokuwikiPage.php [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+ * @author Dominik Brader <support@findologic.com>
+ */
+
+class DokuwikiPage
+{
+    /**
+     * This is the time format as it gets outputted to the template.
+     * DAY.MONTH YEAR (HOUR:MINUTE)
+     */
+    const TIME_FORMAT = '%d.%B %Y (%H:%M)';
```
Comment (georgms) [inline DokuwikiPage.php]: Format in template, use `localizeddate` Twig extension.

File: DokuwikiPage.php [hunk 33]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * @var string Author of the last change that was made to the DokuWiki page.
+     */
+    public $author;
+
+    /**
+     * @var string Formatted timestamp of last edited timestamp.
+     */
+    public $lastEdit;
```
Comment (georgms) [inline DokuwikiPage.php]: Save time as `DateTime` object, this provides type-safety, object-oriented coding and can be properly compared to other `DateTime`s.

File: PageGetter.php [hunk 34]
```diff
@@ -0,0 +1,44 @@
...
+        }, ARRAY_FILTER_USE_BOTH);
+
+        $sortedPages = array_values($allPagesWithoutTitle); // Sort pages so they start with array key [0]
+
+        foreach ($sortedPages as $key => $sortedPage) {
+            $pagesData[] = new DokuwikiPage($sortedPage);
+        }
+
+        if ($pagesData) {
+            return $pagesData;
```
Comment (georgms) [inline PageGetter.php]: Sort descending by last edit, try to use an anonymous function for this. Maybe there is a shortcut to sort objects by a field value.

File: _test/template.response.test.php [hunk 35]
```diff
@@ -0,0 +1,257 @@
...
+        $legend = $dom->getElementsByTagName('legend')[0];
+        $legend = $legend->nodeValue;
+
+        $expectedLegend = 'Pages without title (2)';
+
+        $this->assertEquals($expectedHeader, $header, 'Expected header should be equal to the translation set in the english translation file.');
+        $this->assertEquals($expectedLegend, $legend, 'Expected legend should be equal to the translation set in the english translation file and should contain one item when one page has no title set.');
+    }
+
+    public function test_response_have_multible_elements_when_six_or_more_pages_has_no_title_set()
```
Comment (georgms) [inline _test/template.response.test.php]: Typo

File: _test/template.response.test.php [hunk 36]
```diff
@@ -0,0 +1,257 @@
...
+
+            $nodes = $finder->query("//td[contains(@class, '$pageField')]");
+
+            for ($i = 0; $i < 5; $i++) {
+                $pageWithoutTitle[$pageField][] = trim($nodes->item($i)->nodeValue);
+            }
+        }
+
+        // page-id
+        $expectedPageWithoutTitle['page-id'][0] = 'noootitle0';
```
Comment (georgms) [inline _test/template.response.test.php]: 1. Build an array with expected page titles
2. Build an array with actual pages titles
3. Sort both by the same criteria (alpha most likely)
4. Compare them with the appropriate assert 

See https://stackoverflow.com/a/12496979/1478499

**Review by georgms — CHANGES_REQUESTED** [review]

File: DokuwikiPage.php (hunk 32 above)
Comment (TheKeymaster) [reply DokuwikiPage.php]: Note4Me: should be done locally

File: DokuwikiPage.php (hunk 33 above)
Comment (TheKeymaster) [reply DokuwikiPage.php]: Note4Me: should be done locally

File: _test/template.response.test.php (hunk 35 above)
Comment (TheKeymaster) [reply _test/template.response.test.php]: Note4Me: should be done locally

File: PageGetter.php (hunk 34 above)
Comment (TheKeymaster) [reply PageGetter.php]: Note4Me: should be done locally

File: _test/template.response.test.php (hunk 36 above)
Comment (TheKeymaster) [reply _test/template.response.test.php]: Note4Me: should be done locally

File: tpl/admin.tpl (hunk 31 above)
Comment (TheKeymaster) [reply tpl/admin.tpl]: Note4Me: should be done locally

**Comment by TheKeymaster** [comment]
I have made a demo Video. 
[video.zip](https://github.com/findologic/dokuwiki-plugin-findologic-xml-export/files/1535865/video.zip)

File: PageGetter.php [hunk 37]
```diff
@@ -0,0 +1,45 @@
...
+        // And only get pages that do not have a title set.
+        $allPagesWithoutTitle = array_filter($allPages, function ($page, $k) {
+            return (p_get_metadata($page)['description'] !== '' && !p_get_metadata($page)['title']);
+        }, ARRAY_FILTER_USE_BOTH);
+
+        foreach ($allPagesWithoutTitle as $key => $sortedPage) {
+            $pagesData[] = new DokuwikiPage($sortedPage);
+        }
+
+        if ($pagesData) {
```
Comment (georgms) [inline PageGetter.php]: No need to check if there are pages, `usort` should be able to deal with an empty array, and this way you need to have only one `return` path which is always nice to reduce the complexity of a method.

File: _test/template.response.test.php [hunk 38]
```diff
@@ -0,0 +1,328 @@
...
+
+        // page-id
+        $expectedPageWithoutTitle['page-id'] = [
+            'noootitle1338',
+            'noootitle1337',
+            'noootitle356',
+            'noootitle1203',
+            'noootitle1784'
+        ];
+        sort($expectedPageWithoutTitle['page-id']);
```
Comment (georgms) [inline _test/template.response.test.php:156]: I assume you want to sort the elements of `$expectedPageWithoutTitle` by `page-id`. I don't think this does what you want:

```
$ php -a
Interactive mode enabled

php > $a = [['key' => 'valueB'], ['key' => 'valueA']];
php > var_dump($a);
php shell code:1:
array(2) {
  [0] =>
  array(1) {
    'key' =>
    string(6) "valueB"
  }
  [1] =>
  array(1) {
    'key' =>
    string(6) "valueA"
  }
}
php > sort($a['key']);
PHP Warning:  sort() expects parameter 1 to be array, null given in php shell code on line 1
PHP Stack trace:
PHP   1. {main}() php shell code:0
PHP   2. sort() php shell code:1
```

Additionally, this should throw a warning which should make your tests fail. Please make sure that they fail.

File: _test/template.response.test.php [hunk 39]
```diff
@@ -0,0 +1,328 @@
...
+
+        // page-url
+        $expectedPageWithoutTitle['page-url'] = [
+            'http://wiki.example.com/./doku.php?id=noootitle1338',
+            'http://wiki.example.com/./doku.php?id=noootitle1337',
+            'http://wiki.example.com/./doku.php?id=noootitle356',
+            'http://wiki.example.com/./doku.php?id=noootitle1203',
+            'http://wiki.example.com/./doku.php?id=noootitle1784'
+        ];
+        sort($expectedPageWithoutTitle['page-url']);
```
Comment (georgms) [inline _test/template.response.test.php:166]: I assume you want to sort the elements of `$expectedPageWithoutTitle` by `page-id`. I don't think this does what you want:

```
$ php -a
Interactive mode enabled

php > $a = [['key' => 'valueB'], ['key' => 'valueA']];
php > var_dump($a);
php shell code:1:
array(2) {
  [0] =>
  array(1) {
    'key' =>
    string(6) "valueB"
  }
  [1] =>
  array(1) {
    'key' =>
    string(6) "valueA"
  }
}
php > sort($a['key']);
PHP Warning:  sort() expects parameter 1 to be array, null given in php shell code on line 1
PHP Stack trace:
PHP   1. {main}() php shell code:0
PHP   2. sort() php shell code:1
```

Additionally, this should throw a warning which should make your tests fail. Please make sure that they fail.

File: _test/template.response.test.php [hunk 40]
```diff
@@ -0,0 +1,328 @@
...
+
+        // page-author
+        $expectedPageWithoutTitle['page-author'] = [
+            '(external edit)',
+            '(external edit)',
+            '(external edit)',
+            '(external edit)',
+            '(external edit)'
+        ];
+        sort($expectedPageWithoutTitle['page-author']);
```
Comment (georgms) [inline _test/template.response.test.php:176]: I assume you want to sort the elements of `$expectedPageWithoutTitle` by `page-id`. I don't think this does what you want:

```
$ php -a
Interactive mode enabled

php > $a = [['key' => 'valueB'], ['key' => 'valueA']];
php > var_dump($a);
php shell code:1:
array(2) {
  [0] =>
  array(1) {
    'key' =>
    string(6) "valueB"
  }
  [1] =>
  array(1) {
    'key' =>
    string(6) "valueA"
  }
}
php > sort($a['key']);
PHP Warning:  sort() expects parameter 1 to be array, null given in php shell code on line 1
PHP Stack trace:
PHP   1. {main}() php shell code:0
PHP   2. sort() php shell code:1
```

Additionally, this should throw a warning which should make your tests fail. Please make sure that they fail.

File: lang/de/lang.php [hunk 41]
```diff
@@ -0,0 +1,25 @@
...
+$lang['pagesWithoutTitle'] = 'Seiten ohne Titel';
+$lang['wikipage'] = 'Wikiseite';
+$lang['url'] = 'URL';
+$lang['lasteditby'] = 'Zuletzt editiert von';
+$lang['lastedited'] = 'Zuletzt editiert';
+$lang['edit'] = 'Editieren';
+$lang['thereAre'] = 'Es gibt ';
+$lang['morePages'] = ' weitere Seite(n), die keinen Titel gesetzt hat/haben.';
+$lang['allPagesHaveATitle'] = 'Alle Seiten haben einen Titel gesetzt! Es muss nichts angepasst werden.';
+$lang['noTitleTooltip'] = 'Sollten Seiten keinen Titel gesetzt haben, kann man diese in der Suche nur über die Kategorien finden. Außerdem haben Seiten die keinen Titel haben auch in der Suchresultatsseite keinen gesetzten Namen. ';
```
Comment (georgms) [inline lang/de/lang.php]: Is this true? Are pages without title being exported at all?

**Review by georgms — CHANGES_REQUESTED** [review]
* Nice video!
* Some minor changes.
* Almost there!

**Review by georgms — CHANGES_REQUESTED** [review]
* Nice video!
* Some minor changes.
* Almost there!

File: _test/template.response.test.php (hunk 40 above)
Comment (TheKeymaster) [reply _test/template.response.test.php:176]: :white_check_mark:  As we discussed together, this test works fine

File: _test/template.response.test.php (hunk 39 above)
Comment (TheKeymaster) [reply _test/template.response.test.php:166]: :white_check_mark: As we discussed together, this test works fine

File: _test/template.response.test.php (hunk 38 above)
Comment (TheKeymaster) [reply _test/template.response.test.php:156]: :white_check_mark: As we discussed together, this test works fine

File: PageGetter.php (hunk 37 above)
Comment (TheKeymaster) [reply PageGetter.php]: Yes but I need to define `$pagesData` somewhere as an array. That was my solution for now.

File: lang/de/lang.php (hunk 41 above)
Comment (TheKeymaster) [reply lang/de/lang.php]: As far as I can tell we always said that they should get exported, but the user should be notified in the admin page.
In `\DokuwikiXMLExport::getPageIds` we only exclude pages that do not have a description:

```php
$pagesAndDeletedPages = array_filter($pagesAndDeletedPages, function ($page, $k) {
    return (p_get_metadata($page)['description'] !== '');
}, ARRAY_FILTER_USE_BOTH);
```

We could change the behavior, but is this how we want the behavior should be? 
Lets discuss this!

File: PageGetter.php (hunk 37 above)
Comment (TheKeymaster) [reply PageGetter.php]: :white_check_mark:

File: lang/de/lang.php (hunk 41 above)
Comment (TheKeymaster) [reply lang/de/lang.php]: As discussed I will implement it as you suggested.

Comment (TheKeymaster) [reply lang/de/lang.php]: @georgms I will only be able to make the commit tomorrow, because I need to fix ~45% of all tests due to this change.

Comment (TheKeymaster) [reply lang/de/lang.php]: :white_check_mark:

**Comment by TheKeymaster** [comment]
@georgms This is fixed now. Could you review this?

**Review by georgms — APPROVED** [review]
Grand Master approved. Rock'n'Roll!

**Comment by TheKeymaster** [comment]
MERGE!!
